cpp/src/unary/math_ops.cu

```diff
@@ -229,89 +226,252 @@ struct DeviceNot {
   }
 };
 
-template <typename T, typename F>
-static std::unique_ptr<cudf::column> launch(cudf::column_view const& input,
-                                            cudf::unary_op op,
-                                            rmm::mr::device_memory_resource* mr,
-                                            cudaStream_t stream)
+template <typename OutputType, typename UFN, typename InputIterator>
+std::unique_ptr<cudf::column> transform_fn(InputIterator begin,
+                                           InputIterator end,
+                                           rmm::device_buffer&& null_mask,
+                                           size_type null_count,
+                                           rmm::mr::device_memory_resource* mr,
+                                           cudaStream_t stream)
+{
+  auto const size = cudf::distance(begin, end);
+
+  std::unique_ptr<cudf::column> output = make_fixed_width_column(
+    data_type{type_to_id<OutputType>()}, size, null_mask, null_count, stream, mr);
+  if (size == 0) return output;
+
+  auto output_view = output->mutable_view();
+  thrust::transform(
+    rmm::exec_policy(stream)->on(stream), begin, end, output_view.begin<OutputType>(), UFN{});
+  return output;
+}
+
+struct access_index_fn {
+  cudf::column_device_view d_dictionary;
+  __device__ size_type operator()(size_type idx)
+  {
+    if (d_dictionary.is_null(idx)) return 0;
+    return static_cast<size_type>(d_dictionary.element<dictionary32>(idx));
+  };
+};
+
+template <typename T, typename UFN>
+std::unique_ptr<cudf::column> transform_fn(cudf::dictionary_column_view const& input,
+                                           rmm::mr::device_memory_resource* mr,
+                                           cudaStream_t stream)
 {
-  return cudf::unary::launcher<T, T, F>::launch(input, op, mr, stream);
+  auto dictionary_view = cudf::column_device_view::create(input.parent(), stream);
+  auto keys_view       = cudf::column_device_view::create(input.keys(), stream);
+  auto dictionary_itr  = thrust::make_permutation_iterator(
+    keys_view->begin<T>(),
+    thrust::make_transform_iterator(thrust::make_counting_iterator<size_type>(0),
+                                    access_index_fn{*dictionary_view}));
```

Comment:
This seems like a common utility we'd want to reuse elsewhere. Maybe something like a `make_dictionary_iterator(dictionary_column_view const& input)`.